README.md
Cloudex
======

[![](https://img.shields.io/hexpm/v/cloudex.svg)](https://hex.pm/packages/cloudex) ![](https://img.shields.io/hexpm/dt/cloudex.svg) ![](https://img.shields.io/hexpm/dw/cloudex.svg) ![](https://img.shields.io/coveralls/smeevil/cloudex.svg) [![](https://img.shields.io/github/issues/smeevil/cloudex.svg)](https://github.com/smeevil/cloudex/issues) [![](https://img.shields.io/github/issues-pr/smeevil/cloudex.svg)](https://github.com/smeevil/cloudex/pulls) ![](https://semaphoreci.com/api/v1/smeevil/cloudex/branches/master/shields_badge.svg)

Cloudex is an Elixir library that can upload image files or urls to Cloudinary.
There is also a [CLI tool](https://github.com/smeevil/cloudex_cli) available.

## Getting started

```elixir
defp deps do
  [
    {:cloudex, "~> 1.3.0"},
  ]
end
```

If you are using Elixir 1.4 onwards, you can skip this step.
The Cloudex app must be started. This can be done by adding :cloudex to
the applications list in your mix.exs file. An example:

```elixir
  def application do
    [applications: [:logger, :cloudex],
    ]
  end
```

## Settings

Cloudex requires the API credentials of your Cloudinary account.
You can define either as ENV settings using the keys :
```CLOUDEX_API_KEY``` ```CLOUDEX_SECRET``` and  ```CLOUDEX_CLOUD_NAME```

or in your config.exs using:

```elixir
config :cloudex,
  api_key: "my-api-key",
  secret: "my-secret",
  cloud_name: "my-cloud-name"
```

[Jason](http://github.com/michalmuskala/jason) is the default json library in Cloudex. You can configure Cloudex to use another library. For example:

```elixir
config :cloudex, :json_library, YourLibraryOfChoice
```

## Uploading
You can upload image files or urls pointing to an image as follows :

### example

For uploading a url :
```elixir
iex> Cloudex.upload("http://example.org/test.jpg")
{:ok, %Cloudex.UploadedImage{...}}
```

For uploading a file :

```elixir
iex> Cloudex.upload("test/assets/test.jpg")
{:ok, %Cloudex.UploadedImage{...}}
```
You can also upload a list of files, urls, or mix by giving upload a list like :

```elixir
iex> Cloudex.upload(["/non/existing/file.jpg", "http://example.org/test.jpg"])
[{:error, "File /non/existing/file.jpg does not exist."}, {:ok, %Cloudex.UploadedImage{...}}]
```

An example of the %Cloudex.UploadedImage{} Struct looks as follows:

```elixir
%Cloudex.UploadedImage{
    bytes: 22659,
    created_at: "2015-11-27T10:02:23Z",
    etag: "dbb5764565c1b77ff049d20fcfd1d41d",
    format: "jpg",
    height: 167,
    original_filename: "test",
    public_id: "i2nruesgu4om3w9mtk1z",
    resource_type: "image",
    secure_url: "https://d1vibqt9pdnk2f.cloudfront.net/image/upload/v1448618543/i2nruesgu4om3w9mtk1z.jpg",
    signature: "77b447746476c82bb4921fdea62a9227c584974b",
    source: "http://example.org/test.jpg",
    tags: [],
    type: "upload",
    url: "http://images.cloudassets.mobi/image/upload/v1448618543/i2nruesgu4om3w9mtk1z.jpg",
    version: 1448618543,
    width: 250
}
```

You can tag uploaded images with strings:

```elixir
# as array
Cloudex.upload("test/assets/test.jpg", %{tags: ["foo", "bar"]})
# as comma-separated string
Cloudex.upload("test/assets/test.jpg", %{tags: "foo,bar"})

# result
{:ok, %Cloudex.UploadedImage{
    bytes: 22659,
    created_at: "2015-11-27T10:02:23Z",
    etag: "dbb5764565c1b77ff049d20fcfd1d41d",
    format: "jpg",
    height: 167,
    original_filename: "test",
    public_id: "i2nruesgu4om3w9mtk1z",
    resource_type: "image",
    secure_url: "https://d1vibqt9pdnk2f.cloudfront.net/image/upload/v1448618543/i2nruesgu4om3w9mtk1z.jpg",
    signature: "77b447746476c82bb4921fdea62a9227c584974b",
    source: "http://example.org/test.jpg",
    tags: ["foo", "bar"],
    type: "upload",
    url: "http://images.cloudassets.mobi/image/upload/v1448618543/i2nruesgu4om3w9mtk1z.jpg",
    version: 1448618543,
    width: 250
}}
```

List of additional parameters you can use with `upload/2`:
http://cloudinary.com/documentation/image_upload_api_reference#upload

## Cloudinary URL generation

This package also provides an helper to generate urls from cloudinary given a public id of the image.
As a second argument you can pass in options to transform your image according via cloudinary.

Current supported options are :
```elixir
  :aspect_ratio
  :border
  :color
  :coulor
  :crop
  :default_image
  :delay
  :density
  :dpr
  :effect
  :fetch_format
  :flags
  :gravity
  :height
  :opacity
  :overlay
  :quality
  :radius
  :transformation
  :underlay
  :width
  :x
  :y
  :zoom
```
### Example
```elixir
Cloudex.Url.for("a_public_id")
"//res.cloudinary.com/my_cloud_name/image/upload/a_public_id"
```

```elixir
Cloudex.Url.for("a_public_id", %{width: 400, height: 300})
"//res.cloudinary.com/my_cloud_name/image/upload/h_300,w_400/a_public_id"
```

```elixir
Cloudex.Url.for("a_public_id", %{crop: "fill", fetch_format: 'auto', flags: 'progressive', width: 300, height: 254, quality: "jpegmini", sign_url: true})
"//res.cloudinary.com/my_cloud_name/image/upload/s--jwB_Ds4w--/c_fill,f_auto,fl_progressive,h_254,q_jpegmini,w_300/a_public_id"
```

You can add multiple effects using Cloudex.Url.for/2, an example would be adding an overlay to your image, using:
```elixir
Cloudex.Url.for("a_public_id", [
  %{border: "5px_solid_rgb:c22c33", radius: 5, crop: "fill", height: 246, width: 470, quality: 80},
  %{overlay: "my_overlay", crop: "scale", gravity: "south_east", width: 128 ,x: 5, y: 15}
])
"//res.cloudinary.com/my_cloud_name/image/upload/bo_5px_solid_rgb:c22c33,c_fill,h_246,q_80,r_5,w_470/c_scale,g_south_east,l_my_overlay,w_128,x_5,y_15/a_public_id"
```
## Deleting images
You can request deletion from cloudinary using ```Cloudex.delete/1``` function where the first argument should be the public id of the image you want to delete.

### example:
```elixir
iex> Cloudex.delete("public-id-1")
{:ok, %Cloudex.DeletedImage{public_id: "public-id-1"}}

iex> Cloudex.delete(["public-id-1", "public-id-2"])
[
  {:ok, %Cloudex.DeletedImage{public_id: "public-id-1"}},
  {:ok, %Cloudex.DeletedImage{public_id: "public-id-2"}}
]
```



## Phoenix helper
If you are using phoenix, you can create a small helper called for example cl_image_tag
Create a file containing the following :

```elixir
defmodule MyApp.CloudexImageHelper do
  import Phoenix.HTML.Tag

  def cl_image_tag(public_id, options \\ []) do
    transformation_options = %{}
    if Keyword.has_key?(options, :transforms) do
      transformation_options = Map.merge(%{}, options[:transforms])
    end

    image_tag_options = Keyword.delete(options, :transforms)

    defaults = [
      src: Cloudex.Url.for(public_id, transformation_options),
      width: picture.width,
      height: picture.height,
      alt: "image with name #{public_id}"
    ]

    attributes = Keyword.merge(defaults, image_tag_options)

    tag(:img, attributes)
  end
end
```

Then in your ```web.ex``` add the following line in the ```def view``` section:

```elixir
import MyApp.CloudexImageHelper
```

You should now be able to use the helper in your views as follows :

```elixir
cl_image_tag(public_id, class: "thumbnail", transforms: %{opacity: "50", quality: "jpegmini", sign_url: true})
```

## Documentation

Documentation can be found at docs/index.html or [online](http://smeevil.github.io/cloudex)

## Copyright and License

Copyright (C) 2004 Sam Hocevar <sam@hocevar.net>

The Cloudex Elixir library is released under the DWTFYW license. See the [LICENSE](./LICENSE.md) file.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

# 1.4.1 (2020-03-01)
  - Thanks @tonnenpinguin for adding options to delete private images like: `Cloudex.delete(public_id, %{resource_type: "image", type: "private"})`

# 1.4.0 (2020-02-27)
  - Thanks @davepersing to Allow S3 paths to be passed to Cloudinary when using a private bucket.
  - Thanks @saicoder for bumping httpoison and timex


## 1.3.0 (2019-06-07)

  - Thanks again @abitdodgy for interexchanging the json library from Poison to Jason and making it configurable out of the box :)

## 1.2.4 (2019-06-05)

  - Thank you @abitdodgy for adding the ability to delete images by prefix

## 1.2.3 (2019-02-25)

  - Thank you @erikreedstrom for removing resource_type from signing

## 1.2.2 (2019-02-21)

  - Thank you @StareIntoTheBeard for adding the video codec option
  - bumped deps

## 1.2.1 (2018-11-09)

  - Thank you @edouardmenayde for bumping timex from 3.3.0 which contained a serious bug
  - Thank you @neenaoffline for bumping the version in the readme :)

## 1.2.0 (2018-08-27)

  - bumped deps
  - Thank you @suranyami for adding video support by allowing passing in resource_type: "video".
  - Thank you @cmtt for adding context support

## 1.1.3 (2018-04-16)

  - bumped deps

## 1.1.2 (2018-02-19)

  - Thanks @kuon for moving the http upload to https :D

## 1.1.1 (2018-02-14)

  - Added moderation to the UploadedImage struct as per request of @gccoding
  - reformatted to elixir 1.6

## 1.1.0 (2018-01-08)

  - Thanks @angelikatyborska for adding Phash support to detect duplicate images

## 1.0.1 (2017-11-27)

  - Thanks @joshsmith for fixing some dialyzer errors and adding a proper type for Cloudex.UploadedImage

## 1.0.0 (2017-10-17)

  - Breaking change : When uploading a single file or url you will no
  longer get a `[{:ok, %Cloudex.UploadedImage{}}]` but just a tuple like
  `{:ok, %Cloudex.UploadedImage{}}`, uploading multiple files/urls will
  return a list of tuples just like it already did.
  - Increased test coverage

## 0.2.2 (2017-10-16)

  - bumped specs

## 0.2.1 (2017-10-06)

  - Enforced strict checking of dialyzer

## 0.2.0 (2017-10-03)

  - minor version bump
  - Clean up and reformatting of code
  - added specs and dialyzed
  - Removed unnecessary complicated api split from live and test
  - Its no longer needed to add `config :cloudex, :cloudinary_api,
  Cloudex.CloudinaryApi` in you config
  - Added VCR test actually test real responses from the live api

## 0.1.20 (2017-08-29)

  - Thanks @Tyler-pierce for adding support to rotate images

## 0.1.19 (2017-07-10)

  - Thanks @nbancajas for fixing signing bug when appending secret

## 0.1.18 (2017-07-05)

  - Thanks @dkln for added the g_face option

## 0.1.17 (2017-04-12)

  - You can add multiple effects using Cloudex.Url.for/2, an example would be adding an overlay to your image, using:

```
Cloudex.Url.for("a_public_id", [
  %{border: "5px_solid_rgb:c22c33", radius: 5, crop: "fill", height: 246, width: 470, quality: 80},
  %{overlay: "my_overlay", crop: "scale", gravity: "south_east", width: 128 ,x: 5, y: 15}
])

"//res.cloudinary.com/my_cloud_name/image/upload/bo_5px_solid_rgb:c22c33,c_fill,h_246,q_80,r_5,w_470/c_scale,g_south_east,l_my_overlay,w_128,x_5,y_15/a_public_id"
```

## 0.1.16 (2017-04-03)

  - Thanks to @sudostack for fixing a bug when not passing a transformation string and using signed_urls
  - Bumped deps

## 0.1.15 (2017-01-16)

  - Thanks to @pauloancheta you can now also delete a list with ```Cloudex.delete/1```

## 0.1.13 (2017-01-09)

  - Thanks to @pauloancheta its now possible to also delete image given their public id using ```Cloudex.delete/1```
  - bumped deps
  - fixed deprecation warnings for elixir 1.4


## 0.1.12 (2016-11-24)

  - @remiq added support for tags. Quite appreciated :)

## 0.1.11 (2016-10-25)

  - @jprincipe added support for video urls. Thank you !
  - Also bumped dependencies.

## 0.1.10 (2016-08-26)

  - @bgeihsgt (Thanks!) added the following changes :
  - Whenever an HTTPoison call failed, it would fail with a match error. This uses with goodness to bubble up request or JSON parsing errors.
  - I also increased the task timeout on upload to 60 seconds.


## 0.1.9 (2016-08-23)

  - Cloudex.Url.for now accepts a version as well, thank you @manukall
  - bumped deps

## 0.1.8 (2016-08-19)

  - Removed the file and url extension check as requested. It is not the responsibility of this lib.

## 0.1.7 (2016-08-17)

  - The file extension regex is now case insensitive

## 0.1.6 (2016-08-16)

  - keys for upload opts (like "folder") need to be binaries. sorting however puts atoms before binaries, so signing did not work in such cases. Thank you @manukall!

## 0.1.5 (2016-08-11)

  - bumped deps

## 0.1.4 (2016-07-13)

  - updated timex dep

## 0.1.3 (2016-07-13)

  - allow uploading images from https urls, thank you @manukall !

## 0.1.2 (2016-06-27)

  - Bumped dependencies

## 0.1.1 (2016-06-21)

  - added timex to the applications list to allow for exrm releases (@manukall)

## 0.1.0 (2016-06-21)

  - You can now pass %{public_id: "my_public_id"} as options to the cloudinary upload, thanks @manukall !

## 0.0.2 (2016-03-24)

  - Added a url generator Cloudex.Url.for(public_id, options)
  - Added example of a Phoenix Helper

## 0.0.1 (2016-03-24)

  - Initial commit


lib/cloudex/url.ex
defmodule Cloudex.Url do
  @moduledoc "A module that helps creating urls to your cloudinary image, which can optionally transform the image as well."

  @base_url "//res.cloudinary.com"

  @doc """
  Given a cloudinary public id string, this will generate an image url of where the image is hosted.
  You can also pass a map with options to apply transformations to the image, for more information see the documentation.

  ## examples :
  An url to the image at its original dimensions and no transformations

      iex> Cloudex.Url.for("a_public_id")
      "//res.cloudinary.com/my_cloud_name/image/upload/a_public_id"

  An url to the image with just a signature

      iex> Cloudex.Url.for("a_public_id", %{sign_url: true})
      "//res.cloudinary.com/my_cloud_name/image/upload/s--MXxhpIBQ--/a_public_id"

  An url to the image adjusted to a specific width and height

      iex> Cloudex.Url.for("a_public_id", %{width: 400, height: 300})
      "//res.cloudinary.com/my_cloud_name/image/upload/h_300,w_400/a_public_id"

  An url to the image using multiple transformation options and a signature

      iex> Cloudex.Url.for("a_public_id", %{crop: "fill", fetch_format: 'auto', flags: 'progressive', width: 300, height: 254, quality: "jpegmini", sign_url: true})
      "//res.cloudinary.com/my_cloud_name/image/upload/s--jwB_Ds4w--/c_fill,f_auto,fl_progressive,h_254,q_jpegmini,w_300/a_public_id"

  An url to the image using a named transformation

      iex> Cloudex.Url.for("a_public_id", %{transformation: "my_transformation"})
      "//res.cloudinary.com/my_cloud_name/image/upload/t_my_transformation/a_public_id"

  An url to a specific version of the image

      iex> Cloudex.Url.for("a_public_id", %{version: 1471959066})
      "//res.cloudinary.com/my_cloud_name/image/upload/v1471959066/a_public_id"

  An url to a specific version of the image adjusted to a specific width and height

      iex> Cloudex.Url.for("a_public_id", %{width: 400, height: 300, version: 1471959066})
      "//res.cloudinary.com/my_cloud_name/image/upload/h_300,w_400/v1471959066/a_public_id"

  An url to the image with the file extension of the requested delivery format for the resource. The resource is delivered in the original uploaded format if the file extension is not included.

      iex> Cloudex.Url.for("a_public_id", %{format: "png"})
      "//res.cloudinary.com/my_cloud_name/image/upload/a_public_id.png"

  An url to the resource type.  If resource type not specified, "image" is the default

      iex> Cloudex.Url.for("a_public_id", %{resource_type: "video"})
      "//res.cloudinary.com/my_cloud_name/video/upload/a_public_id"

  An url to the resource type with a named transformation.

      iex> Cloudex.Url.for("a_public_id", %{resource_type: "video", transformation: "my_transformation"})
      "//res.cloudinary.com/my_cloud_name/video/upload/t_my_transformation/a_public_id"

  An url with an overlay
      iex> Cloudex.Url.for("a_public_id", [
      ...>   %{border: "5px_solid_rgb:c22c33", radius: 5, crop: "fill", height: 246, width: 470, quality: 80},
      ...>   %{overlay: "my_overlay", crop: "scale", gravity: "south_east", width: 128 ,x: 5, y: 15}
      ...> ])
      "//res.cloudinary.com/my_cloud_name/image/upload/bo_5px_solid_rgb:c22c33,c_fill,h_246,q_80,r_5,w_470/c_scale,g_south_east,l_my_overlay,w_128,x_5,y_15/a_public_id"

  An url with a face
      iex> Cloudex.Url.for("a_public_id", %{width: 400, height: 300, face: true})
      "//res.cloudinary.com/my_cloud_name/image/upload/g_face,h_300,w_400/a_public_id"

  An url with a video codec setting
      iex> Cloudex.Url.for("a_public_id", %{resource_type: "video", video_codec: "h265"})
      "//res.cloudinary.com/my_cloud_name/video/upload/vc_h265/a_public_id"

  An url with zoom applied to a face

      iex> Cloudex.Url.for("a_public_id", %{zoom: 1.3, face: true, crop: "crop", version: 1471959066})
      "//res.cloudinary.com/my_cloud_name/image/upload/c_crop,g_face,z_1.3/v1471959066/a_public_id"

  An url retaining aspect ratio

      iex> Cloudex.Url.for("a_public_id", %{aspect_ratio: 2.5, width: 400, height: 300, version: 1471959066})
      "//res.cloudinary.com/my_cloud_name/image/upload/ar_2.5,h_300,w_400/v1471959066/a_public_id"
  """

  @spec for(String.t()) :: String.t()
  @spec for(String.t(), map) :: String.t()
  def for(public_id, options \\ %{}) do
    transformations = transformation_string_from(options)

    [
      base_url(),
      resource_type(options),
      "upload",
      signature_for(public_id, options, transformations),
      if(String.length(transformations) > 0, do: transformations),
      version_for(options),
      public_id
    ]
    |> Enum.reject(&(&1 == nil))
    |> Enum.join("/")
    |> append_format(options)
  end

  defp base_url, do: "#{@base_url}/#{Cloudex.Settings.get(:cloud_name)}"

  defp append_format(url, options), do: url <> format(options)

  defp signature_for(public_id, %{sign_url: true}, transformations) do
    to_sign = transformations <> "/#{public_id}" <> Cloudex.Settings.get(:secret)

    signature =
      :sha
      |> :crypto.hash(to_sign)
      |> Base.encode64()
      |> String.slice(0..7)
      |> String.replace("+", "-")
      |> String.replace("/", "_")

    "s--" <> signature <> "--"
  end

  defp signature_for(_, _, _), do: nil

  defp version_for(%{version: version}) when is_integer(version), do: "v#{version}"
  defp version_for(_), do: nil

  defp resource_type(%{resource_type: resource_type}), do: to_string(resource_type)
  defp resource_type(_), do: "image"

  defp format(%{format: format}), do: ".#{format}"
  defp format(_), do: ""

  defp transformation_string_from(options) when is_list(options) do
    options
    |> Enum.map(&transformation_string_from/1)
    |> Enum.join("/")
  end

  defp transformation_string_from(options) when is_map(options) do
    options
    |> Enum.sort()
    |> process()
    |> join_transformations()
  end

  defp join_transformations([]), do: ""
  defp join_transformations(transformations), do: Enum.join(transformations, ",")

  defp process(options) do
    Enum.reduce(options, [], fn {key, value}, acc ->
      acc ++ process_option(key, value)
    end)
  end

  defp process_option(:width, value), do: ["w_#{value}"]
  defp process_option(:height, value), do: ["h_#{value}"]
  defp process_option(:crop, value), do: ["c_#{value}"]
  defp process_option(:aspect_ratio, value), do: ["ar_#{value}"]
  defp process_option(:gravity, value), do: ["g_#{value}"]
  defp process_option(:zoom, value), do: ["z_#{value}"]
  defp process_option(:x, value), do: ["x_#{value}"]
  defp process_option(:y, value), do: ["y_#{value}"]
  defp process_option(:fetch_format, value), do: ["f_#{value}"]
  defp process_option(:quality, value), do: ["q_#{value}"]
  defp process_option(:radius, value), do: ["r_#{value}"]
  defp process_option(:effect, value), do: ["e_#{value}"]
  defp process_option(:opacity, value), do: ["o_#{value}"]
  defp process_option(:border, value), do: ["bo_#{value}"]
  defp process_option(:overlay, value), do: ["l_#{value}"]
  defp process_option(:underlay, value), do: ["u_#{value}"]
  defp process_option(:default_image, value), do: ["d_#{value}"]
  defp process_option(:delay, value), do: ["dl_#{value}"]
  defp process_option(:color, value), do: ["c_#{value}"]
  defp process_option(:coulor, value), do: ["c_#{value}"]
  defp process_option(:dpr, value), do: ["dpr_#{value}"]
  defp process_option(:density, value), do: ["dn_#{value}"]
  defp process_option(:flags, value), do: ["fl_#{value}"]
  defp process_option(:transformation, value), do: ["t_#{value}"]
  defp process_option(:rotation, value), do: ["a_#{value}"]
  defp process_option(:video_codec, value), do: ["vc_#{value}"]
  defp process_option(:face, true), do: ["g_face"]
  defp process_option(_, _), do: []
end
